camel/synthetic_datagen/self_instruct/filter/filter_function.py

```diff
@@ -0,0 +1,158 @@
+
+    Args:
+        min_len (int): The minimum word count required for an instruction.
+        max_len (int): The maximum word count allowed for an instruction.
+    """
+
+    def __init__(self, min_len: int = 5, max_len: int = 200):
+        self.min_len = min_len
+        self.max_len = max_len
+
+    def apply(self, instruction: str) -> bool:
+        r"""Filter the instruction
+
+        Args:
+            instruction (str): the instruction to be filtered
+
+        Returns:
+            bool: True if the length of the instruction is within the range
+            of [min_len, max_len]
+        """
+        word_count = len(instruction.split())
+        return self.min_len <= word_count <= self.max_len
+
+
+class KeywordFilter(FilterFunction):
+    r"""Filters instructions that contain specific undesirable keywords.
+
+    Args:
+        keywords (List[str]): A list of keywords to filter out.
+    """
+
+    def __init__(self, keywords: List[str]):
+        self.keywords = [keyword.lower() for keyword in keywords]
+
+    def apply(self, instruction: str) -> bool:
+        r"""Filter the instruction
+
+        Args:
+            instruction (str): the instruction to be filtered
+
+        Returns:
+            bool: True Instruction must NOT contain any of the keywords.
+        """
+        lower_instr = instruction.lower()
+        return not any(keyword in lower_instr for keyword in self.keywords)
+
+
+class PunctuationFilter(FilterFunction):
+    r"""Filters instructions that begin with a non-alphanumeric character."""
+
+    def apply(self, instruction: str) -> bool:
+        r"""Filter the instruction
+
+        Args:
+            instruction (str): the instruction to be filtered
+
+        Returns:
+            bool: True if the instruction does not start with punctuation.
+        """
+        return not re.match(r'^[^\w\s]', instruction)
+
+
+class NonEnglishFilter(FilterFunction):
+    r"""Filters instructions that do not begin with English letters."""
+
+    def apply(self, instruction: str) -> bool:
+        r"""Filter the instruction
+
+        Args:
+            instruction (str): the instruction to be filtered
+
+        Returns:
+            bool: True if the instruction starts with an English letter.
+        """
+        return bool(re.match(r'^[A-Za-z]', instruction))
+
+
+class RougeSimilarityFilter(FilterFunction):
+    r"""Filters instructions that are too similar to existing instructions
+    based on ROUGE scores.
+
+    Args:
+        existing_instructions (List[str]): A list of existing instructions to
+            compare against.
+        threshold (float): The similarity threshold for filtering (default is
+            0.7).
+    """
+
+    def __init__(
+        self, existing_instructions: List[str], threshold: float = 0.7
+    ):
+        self.existing_instructions = existing_instructions
+        self.threshold = threshold
+        self.rouge = Rouge()
+
+    def apply(self, instruction: str) -> bool:
```

Comment:
Regarding caching, could you share a bit more about how you envision it working in this context? My current understanding is that since each new instruction introduces unique comparisons, it might be challenging to pre-compute and cache the similarities. If we really want to optimize this, I'm thinking vectorizing the instructions and use batch similarity calculations to speed things up. But I might need more time to do some research first.

Reply:
How many items do we expect to possibly be going through here? some vector db techniques can be used but I guess its premature optimisation to do anything like that since there are no simple wins here I think

Reply:
i think for here using vector db could be too heavy, how about
```
class RougeSimilarityFilter(FilterFunction):
    r"""Filters instructions that are too similar to existing instructions
    based on ROUGE scores.

    Args:
        existing_instructions (List[str]): A list of existing instructions to
            compare against.
        threshold (float): The similarity threshold for filtering (default is
            0.7).
    """

    def __init__(
        self, existing_instructions: List[str], threshold: float = 0.7
    ):
        self.existing_instructions = existing_instructions
        self.threshold = threshold
        self.rouge = Rouge()
        self.similarity_cache = {}

    def apply(self, instruction: str) -> bool:
        r"""Filter the instruction

        Args:
            instruction (str): the instruction to be filtered

        Returns:
            bool: True if the instruction's similarity to any existing
                instruction is below the threshold.
        """
        if not self.existing_instructions:
            return True

        for existing_instr in self.existing_instructions:
            # Check if similarity is already cached
            cache_key = tuple(sorted([instruction, existing_instr]))
            if cache_key in self.similarity_cache:
                score = self.similarity_cache[cache_key]
            else:
                # Compute similarity and cache it
                scores = self.rouge.get_scores(instruction, existing_instr)
                score = scores[0]['rouge-l']['f']
                self.similarity_cache[cache_key] = score

            if score > self.threshold:
                return False

        return True
```